README
dircproxy: Detachable IRC Proxy Server
-------------------------------------

dircproxy is an IRC proxy server ("bouncer") designed for people
who use IRC from lots of different workstations or clients, but wish
to remain connected and see what they missed while they were away.
You connect to IRC through dircproxy, and it keeps you connected to
the server, even after you detach your client from it.  While you're
detached, it logs channel and private messages as well as important
events, and when you re-attach it'll let you know what you missed.

This can be used to give you roughly the same functionality as
using ircII and screen together, except you can use whatever IRC
client you like, including X ones!


Features
========

 o Runs on console, as a daemon or from inetd.
 o Able to proxy many simultaneous users and IRC connections.
 o Uses IRC server passwords to authenticate.
 o Remains connected to server when you detach.  To reattach you just
   use the IRC server password again, no special commands!
 o Completely non-blocking throughout.
 o Can connect to servers that also require a password.
 o Can have a list of servers on the same network to connect to, it
   will cycle this list.
 o Throttles data sent to server to ensure you are never flooded
   off.
 o Can check servers to make sure they don't become "stoned".
 o Reconnects to servers if connection is dropped.
 o Can automatically join channels for you on first attach.
 o Rejoins channels if you are kicked off.
 o Can leave channels when you detach and rejoin when you come back.
 o Can take measures to ensure you don't appear "idle" on IRC.
 o Drop's unwanted user modes when you detached, if you forget to
   de-opper yourself.
 o Will refuse to connect to servers that set certain modes
   such as +r.
 o Can bind to any IP on your host to change your appearance on IRC.
 o Can change what username is presented on IRC without affecting
   other users its proxying for.
 o Can send a message to all channels to indicate when you detach
   and reattach.
 o Can change your nickname when you detach.
 o Sets you /AWAY when you detach, if you forget to do so.
 o Fully configurable logging support so when you reattach you
   can see what you missed.
 o Can limit the size of log files to avoid eating diskspace.
 o Log text recalled to your client is sent so your client sees
   it as ordinary IRC text.
 o Can timestamp text in log files so you know when it was sent,
   also can adjust the timestamp you see depending on how long ago
   it was sent.
 o Can make permanent log files for your own use of everything
   on channels or all private messages.
 o Can pass log text to a program of your choice (to send you
   an SMS for example).
 o Can adjust log timestamps depending on timezone difference
   between you and dircproxy.
 o Can proxy DCC chat and sends through itself.
 o Can capture DCC sends and store them on the dircproxy machine
   while you're detached, and even while you're attached.
 o Captured DCC sends can be made subject to a size limit and have
   the sender's nickname included in the filename.
 o Can tunnel DCC sends and chats through ssh tunnels.
   See the included README.dcc-via-ssh for more information.
 o Customisable message of the day for users which can include
   stats about log file sizes.
 o /DIRCPROXY command interface for users to do extra things with
   the proxy.  Fully documented through /DIRCPROXY HELP command, and
   the admin can enable and disable any command on
   a user-by-user basis.
 o Host and password based security.
 o Easy to configure and get running.

All of dircproxy's features are completely configurable and can be
enabled, disabled and adjusted through the configuration file.


Installation
============

See the file INSTALL for building and installation instructions.


Running dircproxy
=================


Once you have installed and configured dircproxy, you can run it
from the console.  It will automatically enter the background
(unless you supply the '-D' parameter) and begin listening for
incoming connections.

You should not need to modify your IRC client or add any
special scripts to support dircproxy.  Authentication is done
using the standard IRC server password, your IRC client should
support this. If you're not sure, try something like '/server
localhost:57000:password' from your IRC client's console.

Connect to it with your IRC client, supplying the same password as
you set in the config file, dircproxy will then connect to the IRC
server for you and begin proxying your session.

When you detach, dircproxy will remain running and remain connected
to the IRC server, logging any channel or private messages for you
until your return.

To re-attach, just connect with your IRC client, again supplying
the password.  dircproxy will see you have left a session running and
re-attach you to that instead of creating a new one.  Messages you
missed while you were away will be sent to you in such a way that
they will fill your IRC windows as if they'd just arrived.


You can also run dircproxy from inetd if you wish, although their
are very few reasons to do this.  See the file README.inetd for
more information on this feature.


Adjusting the Proxy
===================

While you are connected to the proxy you may perform a number of
actions to adjust your proxy session using the '/dircproxy' command.
For example, to end your proxy session, disconnecting you from the
server, use '/dircproxy quit'.  For more information on what other
commands are available, see '/dircproxy help'.


More Information
================

The dircproxy home page is at:

	http://www.dircproxy.net/


Please submit bug reports at:

	http://bugzilla.dircproxy.net/

Also join us on the #dircproxy IRC channel on irc.openprojects.net.


dircproxy is distributed according to the GNU General Public License.
See the file COPYING for details.

Copyright (C) 2000,2001,2002,2003 Scott James Remnant <scott@netsplit.com>.


NEWS
NEWS about changes for existing users.
--------------------------------------

Version 1.2.0-RC1 - 09 jan 2009
	* Lot of bugfixes

Version 1.2.0-beta - 25 Oct 2005 
        * Nickserv support
	* Binding to a local IP
	* a new /dircproxy RELOAD to reload the config
	* Add DCC Resume
	* Add 005 Support
	* Support for IPV6 (experimental)
	* a new /dircproxy NOTIFY command	
	* Support for MD5 password

Version 1.1.0 - Sun Aug 18 02:45:00 BST 2002

	* New 'log_timestamp' config directive.
	* Removed 'chan_log_timestamp' and 'other_log_timestamp'
	  config directives.
	* New 'log_relativetime' config directive.
	* Removed 'chan_log_relativetime' and 'other_log_relativetime'
	  config directives.
	* New 'log_dir' config directive.
	* Removed 'chan_log_copydir' and 'other_log_copydir' config directives.
	* New 'log_program' config directive.
	* Removed 'chan_log_program' and 'other_log_program' config directives.
	* New 'server_log_enabled', 'server_log_always', 'server_log_maxlines'
	  and 'server_log_recall' config directives.
	* Removed 'other_log_enabled', 'other_log_always', 'other_log_maxlines'
	  and 'other_log_recall' config directives.
	* Private messages separated from server messages, new directives
	  'private_log_enabled', 'private_log_always', 'private_log_maxlines'
	  and 'private_log_recall'
	* Log file format created by 'log_dir' no longer has '@1234' on the
	  front if 'log_relativetime' is 'yes'

Version 1.0.3 - Fri Feb  8 17:45:00 GMT 2002

	* New 'pid_file', 'allow_users' and 'allow_kill' config
	  directives.
	* New -p command line parameter (adjusts pid_file).
	* Included contrib scripts, cronchk.sh and privmsg-log.pl

Version 1.0.2 - Tue Jan  1 18:15:00 GMT 2001

	* New 'server_keepalive' config directive.

Version 1.0.1 - Wed Jul 12 17:00:00 BST 2001

	* Just bug fixes.  No changes to worry about.

Version 1.0.0 - Thu Jan 11 15:20:00 GMT 2001

	* Just bug fixes.  No changes to worry about.

Version 0.99.0 - Tue Dec 26 17:55:00 GMT 2000

	* New 'nick_keep' config directive.  The nickname code has been
	  reworked to accommodate this, as such if the connecting client
	  has a different nickname to dircproxy, the connecting client is
	  now corrected (not dircproxy, as it used to be).

	* New 'ctcp_replies' config directive.

Version 0.8.4 - Thu Dec  7 17:35:00 GMT 2000

	* New 'initial_modes' config directive.

Version 0.8.3 - Thu Nov 24 13:45:00 GMT 2000

	* Just bug fixes.  No changes to worry about.

Version 0.8.2 - Thu Nov 23 12:30:00 GMT 2000

	* Just bug fixes.  No changes to worry about.

Version 0.8.1 - Wed Nov 15 16:25:00 GMT 2000

	* Just bug fixes.  No changes to worry about.

Version 0.8.0 - Fri Nov 10 15:20:00 GMT 2000

	* 'chan_log_dir' and 'other_log_dir' have been removed because
	  they cause usage and security problems.  These have been replaced
	  with 'chan_log_copydir' and 'other_log_copydir' which have a
	  slightly different purpose.

	* The log file format on disk has changed slightly to allow for the
	  new 'relative timestamping'.

	* DCC chat's and send's are now proxied by default through dircproxy.

	* Removed config file options:
		server_dnsretry

	* New config file options:
		client_timeout
		connect_timeout
		dns_timeout
		server_throttle
		server_autoconnect
		channel_leave_on_detach
		channel_rejoin_on_attach
		idle_maxtime
		refuse_modes
		quit_message
		attach_message
		detach_message
		chan_log_enabled
		chan_log_relativetime
		chan_log_program
		other_log_enabled
		other_log_relativetime
		other_log_program
		log_timeoffset
		log_events
		dcc_proxy_incoming
		dcc_proxy_outgoing
		dcc_proxy_ports
		dcc_proxy_timeout
		dcc_proxy_sendreject
		dcc_send_fast
		dcc_capture_directory
		dcc_capture_always
		dcc_capture_withnick
		dcc_capture_maxsize
		dcc_tunnel_incoming
		dcc_tunnel_outgoing
		switch_user
		motd_file
		allow_persist
		allow_jump
		allow_jump_new
		allow_host
		allow_die
		
Version 0.7.3 - Wed Sep 27 17:55:00 BST 2000

	* New config file option - "disconnect_on_detach"

Version 0.7.2 - Mon Sep 25 13:13:00 BST 2000

	* If you use a .dircproxyrc in your home directory, this must
	  now have no more then 0700 permissions.  If this is not the
	  case then dircproxy will not start.  This does not apply to
	  the global config file or any supplied with the -f parameter.

Version 0.7.1 - Thu Sep 14 13:15:00 BST 2000

	* Just bug fixes.  No changes to worry about.

Version 0.7.0 - Fri Sep  1 15:00:00 BST 2000

	* The global configuration file is now called dircproxyrc not
	  dircproxy.conf

	* dircproxy is now installed in 'bin' not 'sbin', its a user
	  program after all.

	* Passwords for connection classes must be encrypted in the
	  configuration file using the crypt() function (or contrib/crypt),
	  if you don't like this then edit src/dircproxy.h and turn it off,
	  but thats silly.

	* /DQUIT is gone, try /DIRCPROXY QUIT instead

	* Try /DIRCPROXY HELP

	* Try "man dircproxy" :-)

Version 0.6.2 - Mon Jun 26 16:58:00 BST 2000

	* Changed the logic behind how configuration files are loaded.
	  If you supply one with -f, that is loaded.  If you don't it
	  first tries ~/.dircproxyrc and then tries etc/dircproxy.conf.
	  This means that the config file is no longer installed, and
	  can be found as dircproxy.conf in share/dircproxy or the conf
	  dir of the source.  Also the -G option is no longer valid.

Version 0.6.1 - Thu May 25 15:04:00 BST 2000

	* Just bug fixes.  No changes to worry about.

Version 0.6.0 - Wed May 24 22:33:00 BST 2000

	* There is now a configuration file.  Copy the dircproxyrc.example
	  from the conf directory to your home directory and call it
	  .dircproxyrc - this file contains all the documentation for
	  configuring dircproxy.

Version 0.5.3 - Sat May 13 16:32:00 BST 2000

	* Many of the TODO_CFG_* options defined in dircproxy.h have now
	  been given documentation and renamed.  You'll need to recheck this
	  file if you've changed them.  The PASS/SERVER/BIND ones are still

src/main.c
/* dircproxy
 * Copyright (C) 2000,2001,2002,2003 Scott James Remnant <scott@netsplit.com>.
 * Copyright (C) 2004, 2005, 2006 Francois Harvey <fharvey + dircproxy at securiweb dot net>
 *
 * main.c
 *  - Program main loop
 *  - Command line handling
 *  - Initialisation and shutdown
 *  - Signal handling
 *  - Debug functions
 * --
 * @(#) $Id: main.c,v 1.57 2004/02/13 23:39:34 bear Exp $
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <pwd.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <signal.h>
#include <syslog.h>
#include <errno.h>

#include <dircproxy.h>
#include "getopt/getopt.h"
#include "sprintf.h"
#include "cfgfile.h"
#include "irc_net.h"
#include "irc_client.h"
#include "irc_server.h"
#include "dcc_net.h"
#include "timers.h"
#include "dns.h"
#include "net.h"

/* forward declarations */
static void _sig_term(int);
static void _sig_hup(int);
static void _sig_child(int);
#ifdef DEBUG_MEMORY
static void _sig_usr(int);
#endif /* DEBUG_MEMORY */
static int _reload_config(void);
static int _print_usage(void);
static int _print_version(void);
static int _print_help(void);

/* This is so "ident" and "what" can query version etc - useful (not) */
const char *rcsid = "@(#) $Id: main.c,v 1.57 2004/02/13 23:39:34 bear Exp $";

/* The name of the program */
static char *progname;

/* whether we went in the background or not */
static int in_background = 0;

/* set to 1 to abort the main loop */
static int stop_poll = 0;

/* set to 1 to reload the configuration file */
static int reload_config = 0;

/* Port we're listening on */
static char *listen_port;

/* File to write our pid to */
static char *pid_file;

/* The configuration file we used */
static char *config_file;

/* Global variables */
struct globalvars g;

/* Long options */
static struct option long_opts[] = {
  { "config-file", 1, NULL, 'f' },
  { "help", 0, NULL, 'h' },
  { "version", 0, NULL, 'v' },
#ifndef DEBUG
  { "no-daemon", 0, NULL, 'D' },
#else /* DEBUG */
  { "daemon", 0, NULL, 'D' },
#endif /* DEBUG */
  { "inetd", 0, NULL, 'I' },
  { "listen-port", 1, NULL, 'P' },
  { "pid-file", 1, NULL, 'p' },
  { 0, 0, 0, 0 }
};

/* Options */
#define GETOPTIONS "f:hvDIP:"

/* We need this */
int main(int argc, char *argv[]) {
  int optc, show_help, show_version, show_usage;
  char *local_file, *cmd_listen_port, *cmd_pid_file;
  int inetd_mode, no_daemon;

  /* Set up some globals */
  progname = argv[0];
  listen_port = x_strdup(DEFAULT_LISTEN_PORT);
  pid_file = (DEFAULT_PID_FILE ? x_strdup(DEFAULT_PID_FILE) : 0);

#ifndef DEBUG
  no_daemon = 0;
#else /* DEBUG */
  no_daemon = 1;
#endif /* DEBUG */
  local_file = cmd_listen_port = cmd_pid_file = 0;
  show_help = show_version = show_usage = inetd_mode = 0;
  while ((optc = getopt_long(argc, argv, GETOPTIONS, long_opts, NULL)) != -1) {
    switch (optc) {
      case 'h':
        show_help = 1;
        break;
      case 'v':
        show_version = 1;
        break;
      case 'D':
#ifndef DEBUG
        no_daemon = 1;
#else /* DEBUG */
        no_daemon = 0;
#endif /* DEBUG */
        break;
      case 'I':
        inetd_mode = 1;
        break;
      case 'P':
        free(cmd_listen_port);
        cmd_listen_port = x_strdup(optarg);
        break;
      case 'p':
        free(cmd_pid_file);
        cmd_pid_file = x_strdup(optarg);
        break;
      case 'f':
        free(local_file);
        local_file = x_strdup(optarg);
        break;
      default:
        show_usage = 1;
        break;
    }
  }

  if (show_usage || (optind < argc)) {
    _print_usage();
    return 1;
  }

  if (show_version) {
    _print_version();
    if (!show_help)
      return 0;
  }

  if (show_help) {
    _print_help();
    return 0;
  }

  /* If no -f was specified use the home directory */
  if (!local_file && !inetd_mode) {
    struct stat statinfo;
    struct passwd *pw;

    pw = getpwuid(geteuid());
    if (pw && pw->pw_dir) {
      local_file = x_sprintf("%s/%s", pw->pw_dir, USER_CONFIG_FILENAME);
      debug("Local config file: %s", local_file);
      if (!stat(local_file, &statinfo) && (statinfo.st_mode & 0077)) {
        fprintf(stderr, "%s: Permissions of %s must be 0700 or "
                        "more restrictive\n", progname, local_file);
        free(local_file);
        return 2;
      }
      if (cfg_read(local_file, &listen_port, &pid_file, &g)) {
        /* If the local one didn't exist, set to 0 so we open
           global one */
        free(local_file);
        local_file = 0;
      } else {
        config_file = x_strdup(local_file);
      }
    }
  } else if (local_file) {
    if (cfg_read(local_file, &listen_port, &pid_file, &g)) {
      /* This is fatal! */
      fprintf(stderr, "%s: Couldn't read configuration from %s: %s\n",
              progname, local_file, strerror(errno));
      free(local_file);
      return 2;
    } else {
      config_file = x_strdup(local_file);
    }
  }

  /* Read global config file if local one not found */
  if (!local_file) {
    char *global_file;

    /* Not fatal if it doesn't exist */
    global_file = x_sprintf("%s/%s", SYSCONFDIR, GLOBAL_CONFIG_FILENAME);
    debug("Global config file: %s", global_file);
    cfg_read(global_file, &listen_port, &pid_file, &g);
    config_file = x_strdup(global_file);
    free(global_file);
  } else {
    free(local_file);
  }

  /* Check we got some connection classes */
  if (!connclasses) {
    fprintf(stderr, "%s: No connection classes have been defined.\n", progname);
    return 2;
  }

  /* -P overrides config file */
  if (cmd_listen_port) {
    free(listen_port);
    listen_port = cmd_listen_port;
  }

  /* -p overrides pid file */
  if (cmd_pid_file) {
    free(pid_file);
    pid_file = cmd_pid_file;
  }

  /* Set signal handlers */
  signal(SIGTERM, _sig_term);
  signal(SIGINT, _sig_term);
  signal(SIGHUP, _sig_hup);
  signal(SIGCHLD, _sig_child);
#ifdef DEBUG_MEMORY
  signal(SIGUSR1, _sig_usr);
  signal(SIGUSR2, _sig_usr);
#endif /* DEBUG_MEMORY */

  /* Broken Pipe?  This means that someone disconnected while we were
     sending stuff.  Naughty! */
  signal(SIGPIPE, SIG_IGN);

  if (!inetd_mode) {
    debug("Ordinary console dodge-monkey mode");

    /* Make listening socket before we fork */
    if (ircnet_listen(listen_port)) {
      fprintf(stderr, "%s: Unable to establish listen port\n", progname);
      return 3;
    }

    /* go daemon here */
    if (!no_daemon) {
      switch (go_daemon()) {
        case -1:
          return -1;
        case 0:
          break;
        default:
          return 0;
      }
    }

  } else {
    /* running under inetd means we are backgrounded right *now* */
    in_background = 1;

    debug("Inetd SuperTed mode!");

    /* Hook STDIN into a new proxy */
    ircnet_hooksocket(STDIN_FILENO);
  }
 
  /* Open a connection to syslog if we're in the background */
  if (in_background)
    openlog(PACKAGE, LOG_PID, LOG_USER);

  if (pid_file) {
    FILE *pidfile;

    pidfile = fopen(pid_file, "w");
    if (pidfile) {
      fchmod(fileno(pidfile), 0600);
      fprintf(pidfile, "%d\n", getpid());
      fclose(pidfile);
    } else {
      syscall_fail("fopen", pid_file, 0);
    }
  }
  
  /* Main loop! */
  while (!stop_poll) {
    int ns, nt, status;
    pid_t pid;

    ircnet_expunge_proxies();
    dccnet_expunge_proxies();
    ns = net_poll();
    nt = timer_poll();

    /* Reap any children */
    while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
      debug("Reaped process %d, exit status %d", pid, status);
      
      /* Handle any DNS children */
      dns_endrequest(pid, status);
    }

    /* Reload the configuration file? */
    if (reload_config) {
      _reload_config();
      reload_config = 0;
    }

    if (!ns && !nt)
      break;
  }

  if (pid_file) {
    unlink(pid_file);
  }

  /* Free up stuff */
  ircnet_flush();
  dccnet_flush();
  dns_flush();
  timer_flush();

  /* Do a lingering close on all sockets */
  net_closeall();
  net_flush();

  /* Close down and free up memory */
  if (!inetd_mode && !no_daemon)
    closelog();
  free(listen_port);
  free(pid_file);
  free(config_file);

#ifdef DEBUG_MEMORY
  mem_report("termination");
#endif /* DEBUG_MEMORY */

  return 0;
}

/* Signal to stop polling */
static void _sig_term(int sig) {
  debug("Received signal %d to stop", sig);
  stop();
}

/* Signal to reload configuration file */
static void _sig_hup(int sig) {
  debug("Received signal %d to reload config", sig);
  reload_config = 1;

  /* Restore the signal */
  signal(sig, _sig_hup);
}

/* Signal to reap child process.  Don't do anything other than interrupt
 * whatever we're doing so we reach the waitpid loop in the main loop quicker
 */
static void _sig_child(int sig) {
  debug("Received signal %d to reap", sig);

  /* Restore the signal */
  signal(sig, _sig_child);
}


#ifdef DEBUG_MEMORY
/* On USR signals, dump debug information */
static void _sig_usr(int sig) {
  mem_report(sig == SIGUSR1 ? 0 : "signal");
  signal(sig, _sig_usr);
}
#endif /* DEBUG_MEMORY */

/* Reload the configuration file */
static int _reload_config(void) {
  struct ircconnclass *oldclasses, *c;
  struct globalvars newglobals;
  char *new_listen_port, *new_pid_file;

  debug("Reloading config from %s", config_file);
  new_listen_port = x_strdup(DEFAULT_LISTEN_PORT);
  new_pid_file = (DEFAULT_PID_FILE ? x_strdup(DEFAULT_PID_FILE) : 0);
  oldclasses = connclasses;
  connclasses = 0;

  if (cfg_read(config_file, &new_listen_port, &new_pid_file, &newglobals)) {
    /* Config file reload failed */
    error("Reload of configuration file %s failed", config_file);
    ircnet_flush_connclasses(&connclasses);

    connclasses = oldclasses;
    free(new_listen_port);
    free(new_pid_file);
    return -1;
  }

  /* Copy over new globals */
  memcpy(&g, &newglobals, sizeof(struct globalvars));

  /* Listen port changed */
  if (strcmp(listen_port, new_listen_port)) {
    debug("Changing listen_port from %s to %s", listen_port, new_listen_port);
    if (ircnet_listen(new_listen_port)) {
      /* This isn't fatal */
      error("Unable to change listen_port from %s to %s",
            listen_port, new_listen_port);
    } else {
      free(listen_port);
      listen_port = new_listen_port;
      new_listen_port = 0;
    }
  }

  /* Change the pid file variable (don't bother rewriting it) */
  free(pid_file);
  pid_file = new_pid_file;
  new_pid_file = 0;
  
  /* Match everything back up to the old stuff */
  c = connclasses;
  while (c) {
    struct ircconnclass *o;

    o = oldclasses;
    while (o) {
      if (!strcmp(c->password, o->password)) {
        struct ircproxy *p;

        p = ircnet_fetchclass(o);
        if (p)
          p->conn_class = c;

        break;
      }

      o = o->next;
    }

    c = c->next;
  }

  /* Kill anyone who got lost in the reload */
  c = oldclasses;
  while (c) {
    struct ircproxy *p;

    p = ircnet_fetchclass(c);
    if (p) {
      p->conn_class = 0;
      ircserver_send_command(p, "QUIT", ":Permission revoked - %s %s",
                             PACKAGE, VERSION);
      ircserver_close_sock(p);

      ircclient_send_error(p, "No longer permitted to use this proxy");
      ircclient_close(p);
    }

    c = c->next;
  }
  
  /* Clean up */
  ircnet_flush_connclasses(&oldclasses);
  free(new_listen_port);

  return 0;
}

/* Print the usage instructions to stderr */
static int _print_usage(void) {
  fprintf(stderr, "%s: Try '%s --help' for more information.\n",
          progname, progname);

  return 0;
}

/* Print the version number to stderr */
static int _print_version(void) {
  fprintf(stderr, "%s %s - (C) 2006 Francois Harvey - http://dircproxy.securiweb.net/", PACKAGE, VERSION);
#ifdef SSL
  fprintf(stderr, " - SSL");
#endif /* SSL */
#ifdef DEBUG
  fprintf(stderr, " - DEBUG");
#endif /* DEBUG */
   fprintf(stderr, "\n");
  return 0;
}

/* Print the help to stderr */
static int _print_help(void) {
  fprintf(stderr, "%s %s.  Detachable IRC proxy.\n", PACKAGE, VERSION);
  fprintf(stderr, "(C) 2006 Francois Harvey - http://dircproxy.securiweb.net/\n\n");
  fprintf(stderr, "Usage: %s [OPTION]...\n\n", progname);
  fprintf(stderr, "If a long option shows an argument as mandatory, then "
                  "it is mandatory\nfor the equivalent short option also.  "
                  "Similarly for optional arguments.\n\n");
  fprintf(stderr, "  -h, --help              Print a summary of the options\n");
  fprintf(stderr, "  -v, --version           Print the version number\n");
#ifndef DEBUG
  fprintf(stderr, "  -D, --no-daemon         Remain in the foreground\n");
#else /* DEBUG */
  fprintf(stderr, "  -D, --daemon            Run as a daemon to debug it\n");
#endif /* DEBUG */
  fprintf(stderr, "  -I, --inetd             Being run from inetd "
                                            "(implies -D)\n");
  fprintf(stderr, "  -P, --listen-port=PORT  Port to listen for clients on\n");
  fprintf(stderr, "  -p, --pid-file=FILE     Write PID to this file\n");
  fprintf(stderr, "  -f, --config-file=FILE  Use this file instead of the "
                                            "default\n\n");

  return 0;
}

/* Called when a system call fails.  Print it to stderr or syslog */
int syscall_fail(const char *function, const char *arg, const char *message) {
  char *msg;
  int err;

  err = errno;

  msg = x_sprintf("%s(%s) failed: %s", function, (arg ? arg : ""),
                  (message ? message : strerror(err)));
  if (in_background) {
    syslog(LOG_NOTICE, "%s", msg);
  } else {
#ifdef DEBUG
    fprintf(stderr, "%s: \033[33;1m%s\033[m\n", progname, msg);
#else /* DEBUG */
    fprintf(stderr, "%s: %s\n", progname, msg);
#endif /* DEBUG */
  }

  free(msg);
  return 0;
}

/* Called to log an error */
int error(const char *format, ...) {
  va_list ap;
  char *msg;

  va_start(ap, format);
  msg = x_vsprintf(format, ap);
  va_end(ap);
 
  if (in_background) {
    syslog(LOG_ERR, "%s", msg);
  } else {
#ifdef DEBUG
    fprintf(stderr, "%s: \033[31;1m%s\033[m\n", progname, msg);
#else /* DEBUG */
    fprintf(stderr, "%s: %s\n", progname, msg);
#endif /* DEBUG */
  }

  free(msg);
  return 0;
}

/* Called to output debugging information to stderr or syslog */
int debug(const char *format, ...) {
#ifdef DEBUG
  va_list ap;
  char *msg;

  va_start(ap, format);
#ifdef DEBUG_MEMORY
  msg = xx_vsprintf(0, 0, format, ap);
#else /* DEBUG_MEMORY */
  msg = x_vsprintf(format, ap);
#endif /* DEBUG_MEMORY */
  va_end(ap);
 
  if (in_background) {
    syslog(LOG_DEBUG, "%s", msg);
  } else {
    printf("%s\n", msg);
  }

  free(msg);
#endif /* DEBUG */

  return 0;
}

/* Called to stop dircproxy */
int stop(void) {
  stop_poll = 1;

  return 0;
}

/* Called to set queue config reload */
int reload(void) {
  reload_config = 1;

  return 0;
}

/* Called to go into the background */
int go_daemon(void) {
  if (in_background)
    return 0;

  debug("Running in the background");

  switch (fork()) {
    case -1:
      syscall_fail("fork", "first", 0);
      return -1;
    case 0:
      break;
    default:
      return 1;
  }

  /* Become process group leader */
  setsid();

  switch (fork()) {
    case -1:
      syscall_fail("fork", "second", 0);
      return -1;
    case 0:
      break;
    default:
      return 2;
  }
  
  /* Set our umask */    
  umask(0);

  /* Okay, we're in the background now */
  in_background = 1;
  return 0;
}
